internal/persistence/error.go

```diff
@@ -8,6 +8,9 @@ import (
 // expected amount. This error should usually not be seen under normal circumstances.
 var ErrInvalidRowsAffected = fmt.Errorf("invalid rows affected")
 
+// ErrUniqueViolation is used to indicate a unique constraint violation error.
+var ErrUniqueViolation = fmt.Errorf("unique constraint violation error")
```

Comment:
(this part of the comment is basically moot due to the last paragraph)

If this were to occur, this error would not be properly returned to callers. We'll need to update the `store.HandleErr()` to check for this & then re-write the error to a public-facing error.

Additionally, should this be before to the `store` package, as that's where the `store.ErrNotFound` error exists today?

Can we get this covered by an integration test?

----

Lastly, why create a new error for this in the first place? Why aren't we using `store.ErrConstraint` instead?

Reply:
Actually, I used the same pattern than we have for the different layers. in the persistence one, errors are mapped to what we find in the `persistence/error.go` file. like `persistence.ErrNotFound`.: https://github.com/Kong/koko/blob/main/internal/persistence/error.go#L12

then, at the higher level (store), they are mapped to the proper `store.Err*` ones: like [here](https://github.com/Kong/koko/blob/main/internal/store/store.go#L302) and [here](https://github.com/Kong/koko/pull/554/files#diff-e229227e6e2d6b98849b281cf747ac752e94e82988e36592273e99528f57cbcaR98)

which is then handled by `HandleErr`

Reply:
To get this covered this by an integration tests, it will need some work to get multiple concurrent transactions because at the moment, KoKo is already checking itself if the key exists then return ErrConstraint.